### PR TITLE
Revamp TreeKEM Test Vector

### PR DESCRIPTION
**Pull request by bifurcation**
This PR proposes a new design for the TreeKEM test vector.  It tests the following processes:

1. Validating that an UpdatePath is parent-hash valid before using it
2. Processing the proper element of an UpdatePath to get a path secret and commit secret 
3. Creating an UpdatePath that can be decrypted by the code validated in the previous step
4. Merging the UpdatePath into the tree

Steps 2 and 3 are performed for all sender-receiver pairs.

### CODE REVIEW

File: test-vectors.md [hunk 1]
```diff
@@ -459,53 +459,87 @@ Verification:
...
+    * Identify the node in the tree with node index `node` in the array
+      representation of the tree
+    * Set the private value at this node based on `path_secret`
+  * Verify that the resulting private state `leaf_private[i]` is consistent with
+    the `ratchet_tree`, in the sense that for every node in the private state,
+    the corresponding node in the tree is (a) not blank and (b) contains the
+    public key corresponding to the private key in the private state.
+* Construct a GroupContext object using the provided `cipher_suite`, `group_id`,
+  `epoch`, and `confirmed_transcript_hash`, and the root tree hash of
+  `ratchet_tree
```
Comment (mulmarta) [inline test-vectors.md:526]:
```suggestion
  `ratchet_tree`
```

File: test-vectors.md [hunk 2]
```diff
@@ -459,53 +459,87 @@ Verification:
...
+          "node": /* uint32 node index (in the array representation of the tree) */,
+          "path_secret": /* hex encoded binary path secret */
+        }
+        ...
+      ]
+    }
+    ...
+  ],
+
+  "paths": [
```
Comment (franziskuskiefer) [inline test-vectors.md:492]:
```suggestion
  "update_paths": [
```

**Comment by mulmarta** [comment]
Trying to verify, I'm getting an HPKE error from `context.Open`. Inputs to the `open` function for the first test case and first path i.e. leaf 0 tries to decrypt the update from leaf 1 are below. I wasn't sure if epoch should be incremented?

secret key: `7957e5368bbfe5dcfcce0cac8868d3d1943d7733f5c77ef582ee84ab6a281176`

context:
~~~
16 // label length
  4d4c5320312e3020557064617465506174684e6f6465 // "MLS 1.0 UpdatePathNode"
4050 // context length
  0001 // version
  0001 // ciphersuite
  00 // group id = ""
  0000000000006d9e // given epoch + 1
  20 // tree hash after
    3540f836dc03b81a08d7f7842d1b160ccfa761f65ef47d504db637b743c33eeb
  20 // transcript hash before
    7977581cbf08ae1819408f7c1feb8c8c0833fe6d68468f278b1902cff981dbb5
  00 // extensions
~~~

ciphertext `d78004b758c906a320346d120c2477c8da698f020c92c85eb7757c1ee31c6653f27c4e7499f91b53a9abe4452c749eab`

kem output `8df1383bbf413c3d73c2de55337499d50a628d2babcf8870971c483dfd159833`

**Comment by bifurcation** [comment]
This looks like a bug in MLSpp.  Right now, the merge operation is separate and done after the TreeKEM encryption.  So the context that goes into the TreeKEM encryption reflects the state of the tree after the proposals have been applied, but before the sender's state has been updated.  I'll update MLSpp and post a new vector.

**Comment by franziskuskiefer** [comment]
I'm still failing on larger trees but that's probably my test or OpenMLS. Generally this looks like it works.

**Comment by mulmarta** [comment]
I'm passing :)

**Pull request merged**